src/optimizer/OptimizerUtils.cpp

```diff
@@ -463,14 +482,399 @@ Value OptimizerUtils::normalizeValue(const meta::cpp2::ColumnDef& col, const Val
+        }
+        if (h.hint.begin_value_ref().is_set()) {
+            const auto& value = h.hint.get_begin_value();
+            if (begin->empty() || *begin < value) {
+                *begin = value;
+            }
+        }
+        if (h.hint.end_value_ref().is_set()) {
+            const auto& value = h.hint.get_end_value();
+            if (end->empty() || *end > value) {
+                *end = value;
+            }
+        }
+    }
+    return !(*begin > *end);
+}
+
+bool getIndexColumnHintInExpr(const ColumnDef& field,
+                              const LogicalExpression* expr,
+                              ScoredColumnHint* hint,
+                              std::vector<Expression*>* operands) {
+    std::vector<ScoredColumnHint> hints;
+    for (auto& operand : expr->operands()) {
+        if (!operand->isRelExpr()) continue;
+        auto relExpr = static_cast<const RelationalExpression*>(operand);
+        auto status = selectRelExprIndex(field, relExpr);
+        if (status.ok()) {
+            hints.emplace_back(std::move(status).value());
+            operands->emplace_back(operand);
+        }
+    }
+
+    if (hints.empty()) return false;
+
+    if (hints.size() == 1) {
+        *hint = hints.front();
+    } else {
+        Value begin, end;
+        if (!mergeRangeColumnHints(hints, &begin, &end)) {
+            return false;
+        }
+        ScoredColumnHint h;
+        h.hint.set_column_name(field.get_name());
+        h.hint.set_scan_type(storage::cpp2::ScanType::RANGE);
+        h.hint.set_begin_value(std::move(begin));
+        h.hint.set_end_value(std::move(end));
+        h.score = IndexScore::kRange;
+        *hint = std::move(h);
+    }
+
+    return true;
+}
+
+std::vector<const Expression*> collectUnusedExpr(
+    const LogicalExpression* expr,
+    const std::unordered_set<const Expression*>& usedOperands) {
+    std::vector<const Expression*> unusedOperands;
+    for (auto& operand : expr->operands()) {
+        auto iter = std::find(usedOperands.begin(), usedOperands.end(), operand);
+        if (iter == usedOperands.end()) {
+            unusedOperands.emplace_back(operand);
+        }
+    }
+    return unusedOperands;
+}
+
+StatusOr<IndexResult> selectLogicalExprIndex(const LogicalExpression* expr,
+                                             const IndexItem& index) {
+    if (expr->kind() != Expression::Kind::kLogicalAnd) {
+        return Status::Error("Invalid expression kind.");
+    }
+    IndexResult result;
+    result.hints.reserve(index.get_fields().size());
+    std::unordered_set<const Expression*> usedOperands;
+    for (auto& field : index.get_fields()) {
+        ScoredColumnHint hint;
+        std::vector<Expression*> operands;
+        if (!getIndexColumnHintInExpr(field, expr, &hint, &operands)) {
+            break;
+        }
+        result.hints.emplace_back(std::move(hint));
+        for (auto op : operands) {
+            usedOperands.insert(op);
+        }
```

Comment:
Could give some comments? If the first column is EQ logic expression, the error will be returned via line 775?

Reply:
EQ expression is relational expression, this case will be handled by `selectRelExprIndex` function.